### PR TITLE
Gtfs time zone handling

### LINKED ISSUE #2348: fct_daily_scheduled_trips should handle feed time zone
Related to but distinct from #2347: we don't currently do anything to handle time zones in `fct_daily_scheduled_trips` (or its parent tables, particularly the calendar-related ones). A feed's time zone is set in `dim_agency`. The most important one here is Amtrak -- at least on paper they set their feed time zone to Eastern Time even for services in California.

In addition to the work in #2347 to bridge between RT & Schedule, we also need handling *within schedule* to regularize to a specific time zone. 

See here from the GTFS schedule spec -- even if a stop's time zone is specified it doesn't trump the agency time zone:
image

### PR DESCRIPTION
**Pull request by lauriemerrell**
# Description

This PR adds timezone localization information to GTFS schedule data:
* Adds feed_timezone in all schedule dimension tables (suppressed from open data publishing)
* Adds `trip_start_timezone` and `trip_end_timezone` to `int_gtfs_schedule__stop_times_grouped` and `fct_daily_scheduled_trips`
* Adds trip-localized and Pacific Time trip start and end times as DATETIMEs in `fct_daily_scheduled_trips`
* Uses `service_date` as the date summary level for stops, shapes, and daily schedule service summary per discussion in Slack with @tiffanychu90 and @edasmalchi 
* Some assorted cleanup: fixes some existing table documentation 
* Fixes daylight savings handling: `service_date` should not be cast directly to a timestamp (i.e. midnight); it is explicitly defined as starting from noon minus 12 hours to handle daylight savings 

Follow-ups:
* Full refresh needed for schedule data (to add the `feed_timezone` field) -- can do immediately after merge: `poetry run dbt run --full-refresh -s dim_schedule_feeds+,config.materialized:incremental --exclude fct_daily_rt_feed_files fct_hourly_rt_feed_files fct_hourly_rt_feed_files_success`
* https://github.com/cal-itp/reports/issues/245
* We initially anticipated a full refresh for RT data but the RT portion of this issue will be addressed in #2489 and the fix for #2471 instead of here (full refresh of #2458 blocked by this collection of work)
* Fix materialization? Cluster fct daily scheduled trips / make stop times grouped incremental?: Migrated to #2488 
* Reports site updates needed due to column name change in organization service summary table: Migrated to https://github.com/cal-itp/reports/issues/245

Resolves #2348

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) -- changes date columns but relevant downstream places have been updated
- [ ] Documentation
- [ ] agencies.yml

## How has this been tested?

Have run & tested affected tables locally. Looked specifically at daylight savings day service to confirm new implementation works.

### CODE REVIEW

**Comment by lauriemerrell** [comment]
Waiting on feedback on which date value to use for `fct daily scheduled stops/shapes`, once I can implement that, this will be good to go

File: warehouse/models/mart/gtfs/dim_stops.sql [hunk 1]
```diff
@@ -41,10 +63,15 @@ dim_stops AS (
...
+        coalesce_missing_ids.stop_id IS NULL AS warning_missing_primary_key,
+        stop_timezone_coalesced,
         _feed_valid_from,
-    FROM make_dim
+        feed_timezone,
+    FROM coalesce_missing_ids
     LEFT JOIN bad_rows
-        USING (base64_url, ts, stop_id)
+        USING (base64_url, ts, non_null_stop_id)
+    LEFT JOIN fill_in_tz
```
Comment (atvaccaro) [inline warehouse/models/mart/gtfs/dim_stops.sql:73]: can you not just select from `fill_in_tz` here?

File: warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml [hunk 2]
```diff
@@ -958,18 +1091,21 @@ models:
           that is being summarized here, and counts may be inflated due to multiple rows with identical identifiers.
   - name: fct_daily_scheduled_stops
     description: |
-      A daily table showing all stops that had trip stop events scheduled on a given date.
+      Each row is a summary of trip activity by stop for a given feed on a given agency-defined `service_date`.
+
+      A stop here is defined within an individual feed so, even if a given stop is served by multiple
+      agencies, the stop will have one row per feed in which it appears.
```
Comment (atvaccaro) [inline warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml:1097]: this is a bit unclear; maybe remove `even if` and the reference to agencies?

**Review by atvaccaro — APPROVED** [review]
comprehensive docs 📖

File: warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml (hunk 2 above)
Comment (lauriemerrell) [reply warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml:1097]: The point here is that the same physical location may appear multiple times in the table if it appears in multiple feeds; we are not doing stop entity resolution across feeds (this is a thorny topic in GTFS world.) I can try to clarify but I think that it's important to note that in some form.

File: warehouse/models/mart/gtfs_schedule_latest/_gtfs_schedule_latest.yml [hunk 3]
```diff
@@ -634,7 +644,14 @@ models:
         description: '{{ doc("gtfs_stops__level_id") }}'
       - name: platform_code
         description: '{{ doc("gtfs_stops__platform_code") }}'
-
+      - *feed_timezone
+        # TODO: would be good to actually publish this, with appropriate documentation, next time we are ready to make a round of updates
+      - name: warning_missing_primary_key
+        meta:
+          publish.ignore: true
+      - name: stop_timezone_coalesced
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/mart/gtfs_schedule_latest/_gtfs_schedule_latest.yml:652]: This is a view on top of a table and there's a not-null test on the table.... Do we want to test again?

File: warehouse/models/mart/gtfs_schedule_latest/_gtfs_schedule_latest.yml [hunk 4]
```diff
@@ -63,6 +63,10 @@ models:
         description: '{{ doc("gtfs_agency__agency_fare_url") }}'
       - name: agency_email
         description: '{{ doc("gtfs_agency__agency_email") }}'
+      - &feed_timezone
+        name: feed_timezone
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/mart/gtfs_schedule_latest/_gtfs_schedule_latest.yml:67]: Same answer as stop timezone. I only included it in this YAML to suppress it from publishing

File: warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml (hunk 2 above)
Comment (lauriemerrell) [reply warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml]: https://github.com/cal-itp/data-infra/pull/2457/commits/fba7014673f92f9016c806b0d44e4125fcfe5f00

File: warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml [hunk 5]
```diff
@@ -550,8 +654,20 @@ models:
       - dbt_utils.expression_is_true:
           expression: "ttl_service_hours >= 0"
     columns:
-      - name: activity_date
-        description: Date on which this service level was present
+      - &service_date_for_summaries
+        name: service_date
+        description: |
+          Agency-defined service date on which this service level was present.
+          This respect's the agency's definition of a service day, which may not correspond to
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/mart/gtfs/_mart_gtfs_fcts.yml]: https://github.com/cal-itp/data-infra/pull/2457/commits/fba7014673f92f9016c806b0d44e4125fcfe5f00

File: warehouse/models/intermediate/gtfs/int_gtfs_schedule__stop_times_grouped.sql [hunk 6]
```diff
@@ -3,19 +3,51 @@
...
+    QUALIFY RANK() OVER (PARTITION BY feed_key, stop_id ORDER BY ct DESC) = 1
+),
 
+stops_times_with_tz AS (
     SELECT
+        dim_stop_times.*,
+        COALESCE(FIRST_VALUE(stop_timezone_coalesced)
+            OVER (PARTITION BY feed_key, trip_id
+                ORDER BY stop_sequence
+                ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING), feed_timezone) AS trip_start_timezone,
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/intermediate/gtfs/int_gtfs_schedule__stop_times_grouped.sql:29]: They include it in [the examples](https://cloud.google.com/bigquery/docs/reference/standard-sql/navigation_functions#first_value) so I included it here. But I just did a test and I think you may not need it if you're using ascending sort with first value but you do it you're using last value with descending sort because: 
image

TLDR, seems safer / fine to include it?

[the remark this answers is not in the document]
Comment (atvaccaro) [reply warehouse/models/intermediate/gtfs/int_gtfs_schedule__stop_times_grouped.sql:29]: Ok cool, that link does specify aggregation but I think it's fine

**Review by atvaccaro — APPROVED** [review]
ty ty

File: warehouse/models/intermediate/gtfs/int_gtfs_schedule__stop_times_grouped.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/intermediate/gtfs/int_gtfs_schedule__stop_times_grouped.sql:29]: Pedant mode the screenshot specifies aggregation functions but the actual link is to the navigation functions page where they do include it, and I manually verified that 

```
WITH finishers AS
 (SELECT 'Sophia Liu' as name,
  TIMESTAMP '2016-10-18 2:51:45' as finish_time,
  'F30-34' as division
  UNION ALL SELECT 'Lisa Stelzner', TIMESTAMP '2016-10-18 2:54:11', 'F35-39'
  UNION ALL SELECT 'Nikki Leith', TIMESTAMP '2016-10-18 2:59:01', 'F30-34'
  UNION ALL SELECT 'Lauren Matthews', TIMESTAMP '2016-10-18 3:01:17', 'F35-39'
  UNION ALL SELECT 'Desiree Berry', TIMESTAMP '2016-10-18 3:05:42', 'F35-39'
  UNION ALL SELECT 'Suzy Slane', TIMESTAMP '2016-10-18 3:06:24', 'F35-39'
  UNION ALL SELECT 'Jen Edwards', TIMESTAMP '2016-10-18 3:06:36', 'F30-34'
  UNION ALL SELECT 'Meghan Lederer', TIMESTAMP '2016-10-18 3:07:41', 'F30-34'
  UNION ALL SELECT 'Carly Forte', TIMESTAMP '2016-10-18 3:08:58', 'F25-29'
  UNION ALL SELECT 'Lauren Reasoner', TIMESTAMP '2016-10-18 3:10:14', 'F30-34')
SELECT name,
  FORMAT_TIMESTAMP('%X', finish_time) AS finish_time,
  division,
  FORMAT_TIMESTAMP('%X', fastest_time) AS fastest_time,
  TIMESTAMP_DIFF(finish_time, fastest_time, SECOND) AS delta_in_seconds
FROM (
  SELECT name,
  finish_time,
  division,
  LAST_VALUE(finish_time)
    OVER (PARTITION BY division ORDER BY finish_time DESC
    ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS fastest_time
  FROM finishers)
```
Returns a different result than 
```
WITH finishers AS
 (SELECT 'Sophia Liu' as name,
  TIMESTAMP '2016-10-18 2:51:45' as finish_time,
  'F30-34' as division
  UNION ALL SELECT 'Lisa Stelzner', TIMESTAMP '2016-10-18 2:54:11', 'F35-39'
  UNION ALL SELECT 'Nikki Leith', TIMESTAMP '2016-10-18 2:59:01', 'F30-34'
  UNION ALL SELECT 'Lauren Matthews', TIMESTAMP '2016-10-18 3:01:17', 'F35-39'
  UNION ALL SELECT 'Desiree Berry', TIMESTAMP '2016-10-18 3:05:42', 'F35-39'
  UNION ALL SELECT 'Suzy Slane', TIMESTAMP '2016-10-18 3:06:24', 'F35-39'
  UNION ALL SELECT 'Jen Edwards', TIMESTAMP '2016-10-18 3:06:36', 'F30-34'
  UNION ALL SELECT 'Meghan Lederer', TIMESTAMP '2016-10-18 3:07:41', 'F30-34'
  UNION ALL SELECT 'Carly Forte', TIMESTAMP '2016-10-18 3:08:58', 'F25-29'
  UNION ALL SELECT 'Lauren Reasoner', TIMESTAMP '2016-10-18 3:10:14', 'F30-34')
SELECT name,
  FORMAT_TIMESTAMP('%X', finish_time) AS finish_time,
  division,
  FORMAT_TIMESTAMP('%X', fastest_time) AS fastest_time,
  TIMESTAMP_DIFF(finish_time, fastest_time, SECOND) AS delta_in_seconds
FROM (
  SELECT name,
  finish_time,
  division,
  LAST_VALUE(finish_time)
    OVER (PARTITION BY division ORDER BY finish_time DESC
) AS fastest_time
  FROM finishers)
```
